File: lib/services/cloud_sync_service.dart
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:path/path.dart' as path;
import 'package:path_provider/path_provider.dart';
import '../models/cloud_file.dart';
import 'cloud_storage_service.dart';

/// Service for syncing Scrivener projects with cloud storage
class CloudSyncService extends ChangeNotifier {
  final CloudStorageService _cloudStorageService;

  CloudSyncService(this._cloudStorageService);

  bool _isSyncing = false;
  double _syncProgress = 0.0;
  String? _syncStatus;
  String? _error;

  bool get isSyncing => _isSyncing;
  double get syncProgress => _syncProgress;
  String? get syncStatus => _syncStatus;
  String? get error => _error;

  /// Download a Scrivener project from cloud storage
  /// Returns the local path where the project was downloaded
  Future<String?> downloadProject(CloudFile projectFolder) async {
    if (!projectFolder.isScrivenerProject) {
      _error = 'Selected file is not a Scrivener project';
      notifyListeners();
      return null;
    }

    _isSyncing = true;
    _syncProgress = 0.0;
    _syncStatus = 'Downloading project...';
    _error = null;
    notifyListeners();

    try {
      // Get local directory for storing projects
      final localDir = await _getLocalProjectsDirectory();
      final projectDir = Directory(path.join(localDir.path, projectFolder.name));

      // Create project directory
      if (await projectDir.exists()) {
        await projectDir.delete(recursive: true);
      }
      await projectDir.create(recursive: true);

      // Download all files from the cloud project
      await _downloadFolderRecursively(
        projectFolder.id,
        projectDir.path,
      );

      _isSyncing = false;
      _syncProgress = 1.0;
      _syncStatus = 'Download complete';
      notifyListeners();

      return projectDir.path;
    } catch (e) {
      _error = 'Failed to download project: $e';
      _isSyncing = false;
      notifyListeners();
      return null;
    }
  }

  /// Upload a Scrivener project to cloud storage
  Future<bool> uploadProject(
    String localProjectPath,
    String? cloudParentFolderId,
  ) async {
    _isSyncing = true;
    _syncProgress = 0.0;
    _syncStatus = 'Uploading project...';
    _error = null;
    notifyListeners();

    try {
      final projectDir = Directory(localProjectPath);
      if (!await projectDir.exists()) {
        throw Exception('Project directory does not exist');
      }

      final projectName = path.basename(localProjectPath);

      // Create project folder in cloud if it doesn't exist
      CloudFile? cloudProjectFolder;
      try {
        cloudProjectFolder = await _cloudStorageService.createFolder(
          name: projectName,
          parentFolderId: cloudParentFolderId,
        );
      } catch (e) {
        // Folder might already exist, try to find it
        final files = await _cloudStorageService.listFiles(
          folderId: cloudParentFolderId,
        );
        cloudProjectFolder = files.firstWhere(
          (f) => f.name == projectName && f.isDirectory,
          orElse: () => throw Exception('Failed to create/find project folder'),
        );
      }

      // Upload all files recursively
      await _uploadFolderRecursively(
        projectDir.path,
        cloudProjectFolder.id,
      );

      _isSyncing = false;
      _syncProgress = 1.0;
      _syncStatus = 'Upload complete';
      notifyListeners();

      return true;
    } catch (e) {
      _error = 'Failed to upload project: $e';
      _isSyncing = false;
      notifyListeners();
      return false;
    }
  }

  /// Download a folder and its contents recursively
  Future<void> _downloadFolderRecursively(
    String cloudFolderId,
    String localPath,
  ) async {
    // List files in cloud folder
    final files = await _cloudStorageService.listFiles(folderId: cloudFolderId);

    for (int i = 0; i < files.length; i++) {
      final file = files[i];
      final localFilePath = path.join(localPath, file.name);

      if (file.isDirectory) {
        // Create local directory and download recursively
        final dir = Directory(localFilePath);
        await dir.create(recursive: true);
        await _downloadFolderRecursively(file.id, localFilePath);
      } else {
        // Download file
        final content = await _cloudStorageService.downloadFile(file.id);
        final localFile = File(localFilePath);
        await localFile.writeAsBytes(content);
      }

      // Update progress
      _syncProgress = (i + 1) / files.length * 0.5; // First half of progress
      notifyListeners();
    }
  }

  /// Upload a folder and its contents recursively
  Future<void> _uploadFolderRecursively(
    String localPath,
    String cloudFolderId,
  ) async {
    final dir = Directory(localPath);
    final entities = await dir.list().toList();

    for (int i = 0; i < entities.length; i++) {
      final entity = entities[i];
      final name = path.basename(entity.path);

      if (entity is Directory) {
        // Create cloud folder and upload recursively
        final cloudFolder = await _cloudStorageService.createFolder(
          name: name,
          parentFolderId: cloudFolderId,
        );
        await _uploadFolderRecursively(entity.path, cloudFolder.id);
      } else if (entity is File) {
        // Upload file
        final content = await entity.readAsBytes();
        await _cloudStorageService.uploadFile(
          name: name,
          content: content,
          parentFolderId: cloudFolderId,
        );
      }

      // Update progress
      _syncProgress = 0.5 + (i + 1) / entities.length * 0.5; // Second half
      notifyListeners();
    }
  }

  /// Get local directory for storing downloaded projects
  Future<Directory> _getLocalProjectsDirectory() async {
    if (kIsWeb) {
      // For web, we can't use file system, would need IndexedDB
      // For now, throw an error
      throw UnsupportedError(
        'Web platform requires different storage strategy',
      );
    }

    final appDir = await getApplicationDocumentsDirectory();
    final projectsDir = Directory(path.join(appDir.path, 'CloudProjects'));

    if (!await projectsDir.exists()) {
      await projectsDir.create(recursive: true);
    }

    return projectsDir;
  }

  /// Sync local changes back to cloud
  Future<bool> syncChanges(String localProjectPath, String cloudProjectId) async {
    _isSyncing = true;
    _syncStatus = 'Syncing changes...';
    notifyListeners();

    try {
      // For now, just re-upload the entire project
      // In the future, could implement delta sync
      await uploadProject(localProjectPath, null);
      return true;
    } catch (e) {
      _error = 'Failed to sync changes: $e';
      notifyListeners();
      return false;
    } finally {
      _isSyncing = false;
      notifyListeners();
    }
  }

  /// Clear error message
  void clearError() {
    _error = null;
    notifyListeners();
  }
}

File: react-native-writr/src/services/CloudSyncService.ts
import RNFS from 'react-native-fs';
import {CloudFile, SyncProgress} from '../types';
import CloudStorageService from './CloudStorageService';

/**
 * Service for syncing Scrivener projects between cloud and local storage
 * Handles recursive upload/download of project directories
 */
export class CloudSyncService {
  private cloudStorageService: typeof CloudStorageService;
  private onProgress?: (progress: SyncProgress) => void;

  constructor(cloudStorageService: typeof CloudStorageService) {
    this.cloudStorageService = cloudStorageService;
  }

  /**
   * Set progress callback
   */
  setProgressCallback(callback: (progress: SyncProgress) => void): void {
    this.onProgress = callback;
  }

  /**
   * Download a Scrivener project from cloud to local storage
   */
  async downloadProject(
    projectFolder: CloudFile,
  ): Promise<string | null> {
    try {
      // Create local directory for cloud projects
      const localDir = `${RNFS.DocumentDirectoryPath}/CloudProjects`;
      await this.ensureDirectoryExists(localDir);

      const projectDir = `${localDir}/${projectFolder.name}`;

      // Download all files recursively
      await this.downloadFolderRecursively(
        projectFolder.id,
        projectFolder.name,
        projectDir,
      );

      return projectDir;
    } catch (error) {
      console.error('Error downloading project:', error);
      return null;
    }
  }

  /**
   * Upload a Scrivener project from local to cloud storage
   */
  async uploadProject(
    localProjectPath: string,
    cloudParentFolderId?: string,
  ): Promise<boolean> {
    try {
      const projectName = localProjectPath.substring(
        localProjectPath.lastIndexOf('/') + 1,
      );

      // Create project folder in cloud
      const cloudProjectFolder = await this.cloudStorageService.createFolder(
        projectName,
        cloudParentFolderId,
      );

      if (!cloudProjectFolder) {
        throw new Error('Failed to create project folder in cloud');
      }

      // Upload all files recursively
      await this.uploadFolderRecursively(
        localProjectPath,
        cloudProjectFolder.id,
      );

      return true;
    } catch (error) {
      console.error('Error uploading project:', error);
      return false;
    }
  }

  /**
   * Download a folder recursively
   */
  private async downloadFolderRecursively(
    cloudFolderId: string,
    folderName: string,
    localPath: string,
  ): Promise<void> {
    // Create local directory
    await this.ensureDirectoryExists(localPath);

    // List files in cloud folder
    const files = await this.cloudStorageService.listFiles(cloudFolderId);

    let processed = 0;
    const total = files.length;

    for (const file of files) {
      this.reportProgress(processed, total, `Downloading ${file.name}...`);

      if (file.isDirectory) {
        // Recursively download subfolder
        const subfolderPath = `${localPath}/${file.name}`;
        await this.downloadFolderRecursively(
          file.id,
          file.name,
          subfolderPath,
        );
      } else {
        // Download file
        const filePath = `${localPath}/${file.name}`;
        await this.cloudStorageService.downloadFile(file.id, filePath);
      }

      processed++;
    }

    this.reportProgress(total, total, `Downloaded ${folderName}`);
  }

  /**
   * Upload a folder recursively
   */
  private async uploadFolderRecursively(
    localPath: string,
    cloudParentFolderId: string,
  ): Promise<void> {
    const items = await RNFS.readDir(localPath);

    let processed = 0;
    const total = items.length;

    for (const item of items) {
      this.reportProgress(processed, total, `Uploading ${item.name}...`);

      if (item.isDirectory()) {
        // Create subfolder in cloud
        const cloudSubfolder = await this.cloudStorageService.createFolder(
          item.name,
          cloudParentFolderId,
        );

        if (cloudSubfolder) {
          // Recursively upload subfolder contents
          await this.uploadFolderRecursively(item.path, cloudSubfolder.id);
        }
      } else {
        // Upload file
        await this.cloudStorageService.uploadFile(
          item.path,
          item.name,
          cloudParentFolderId,
        );
      }

      processed++;
    }

    this.reportProgress(total, total, 'Upload complete');
  }

  /**
   * Ensure directory exists, create if not
   */
  private async ensureDirectoryExists(path: string): Promise<void> {
    const exists = await RNFS.exists(path);
    if (!exists) {
      await RNFS.mkdir(path);
    }
  }

  /**
   * Report progress
   */
  private reportProgress(
    current: number,
    total: number,
    message: string,
  ): void {
    if (this.onProgress) {
      this.onProgress({current, total, message});
    }
  }
}

// Export a singleton instance
export default new CloudSyncService(CloudStorageService);
